### PR TITLE
Arbitrary self types v2: (unused) Receiver trait

### PR DESCRIPTION
**Pull request by adetaylor**
This commit contains a new `Receiver` trait, which is the basis for the Arbitrary Self Types v2 RFC. This allows smart pointers to be method receivers even if they're not Deref.

This is currently unused by the compiler - a subsequent PR will start to use this for method resolution if the `arbitrary_self_types` feature gate is enabled. This is being landed first simply to make review simpler: if people feel this should all be in an atomic PR let me know.

This is a part of the arbitrary self types v2 project, https://github.com/rust-lang/rfcs/pull/3519
https://github.com/rust-lang/rust/issues/44874

r? @wesleywiser

### CODE REVIEW

**Comment by adetaylor** [comment]
@wesleywiser, thanks for triggering the crater run on the previous PR. There are [only two crates broken by the removal of the `Receiver` trait](https://github.com/rust-lang/rust/pull/130225#issuecomment-2466182930) - so I think we're good to go ahead and land this PR once you're happy with it.

**Review by wesleywiser — APPROVED** [review]
Changes look good. I don't think this needs a signoff from T-libs for the same reasons as mentioned in https://github.com/rust-lang/rust/pull/130225#issuecomment-2413035918 by T-libs member. I will just cc @rust-lang/libs in case I'm totally off-base here.

**Pull request merged**
